crates/prover-services/src/parallel/mod.rs

```diff
@@ -234,22 +243,38 @@ fn make_proof<Da, Vm, Stf>(
     mut vm: Vm,
     elf: Vec<u8>,
     zk_storage: Stf::PreState,
-    proof_mode: Arc<Mutex<ProofGenMode<Da, Vm, Stf>>>,
+    proof_mode: ProofGenMode<Da, Vm, Stf>,
 ) -> Result<Proof, anyhow::Error>
 where
     Da: DaService,
     Vm: ZkvmHost,
     Stf: StateTransitionFunction<Da::Spec> + Send + Sync,
     Stf::PreState: Send + Sync,
 {
-    let mut proof_mode = proof_mode.blocking_lock();
-    match proof_mode.deref_mut() {
+    match proof_mode {
         ProofGenMode::Skip => Ok(Vec::default()),
-        ProofGenMode::Simulate(ref mut verifier) => verifier
-            .run_sequencer_commitments_in_da_slot(vm.simulate_with_hints(), zk_storage)
-            .map(|_| Vec::default())
-            .map_err(|e| anyhow::anyhow!("Guest execution must succeed but failed with {:?}", e)),
+        ProofGenMode::Simulate(verifier) => {
+            let mut verifier = verifier.blocking_lock();
+            verifier
+                .run_sequencer_commitments_in_da_slot(vm.simulate_with_hints(), zk_storage)
+                .map(|_| Vec::default())
+                .map_err(|e| {
+                    anyhow::anyhow!("Guest execution must succeed but failed with {:?}", e)
+                })
+        }
         ProofGenMode::Execute => vm.run(elf, false),
-        ProofGenMode::Prove => vm.run(elf, true),
+        ProofGenMode::ProveWithSampling => {
+            // `make_proof` is called with a probability in this case.
+            // When it's called, we have to produce a real proof.
+            vm.run(elf, true)
+        }
+        ProofGenMode::ProveWithSamplingWithFakeProofs(proof_sampling_number) => {
```

Comment:
Only nit I have is that ProveWithSampling is a bit confusing to me, because in the eyes of prover it is just Prove, and sampling happens above in the call stack. So i think we should just call them Prove and ProveWithSampling respectively, so its obvious from the ProofGenMode whether it should do any sampling itself.

I am not going to block the PR for this nit, but would love to see this change.

Reply:
Also, what if instead of sampling on per proof basis for ProveWithSamplingWithFakeProofs, we sample per DA block. Meaning, we again decide the sampling on da_block_handler level, but what we decide is to whether proofs in current DA should be run in execute or prove mode. It would simplify the flow quite a bit. Just throwing this out as a thought. WDYT?